src/compute-client/src/controller/instance.rs

```diff
@@ -2193,74 +2308,63 @@ impl<T> CollectionIntrospection<T> {
+            self.send(IntrospectionType::ComputeMaterializedViewRefreshes, updates);
+        }
     }
+}
 
-    /// Should be called whenever the write frontier of the collection advances. It updates the
-    /// state that should be recorded in introspection relations, and sends it through the
-    /// `introspection_tx` channel.
-    fn frontier_update(&mut self, write_frontier: &Antichain<T>) {
-        if let Some(refresh_introspection_state) = &mut self.refresh_introspection_state {
-            // Old row to retract based on old state.
-            let retraction = refresh_introspection_state.row_for_collection(self.collection_id);
-            // Update state.
-            refresh_introspection_state.frontier_update(write_frontier);
-            // New row to insert based on new state.
-            let insertion = refresh_introspection_state.row_for_collection(self.collection_id);
-            // Send changes.
-            self.send(
-                IntrospectionType::ComputeMaterializedViewRefreshes,
-                vec![(retraction, -1), (insertion, 1)],
-            );
+#[derive(Debug)]
+struct FrontiersIntrospectionState<T> {
+    read_frontier: Antichain<T>,
+    write_frontier: Antichain<T>,
+}
+
+impl<T: ComputeControllerTimestamp> FrontiersIntrospectionState<T> {
+    fn new(as_of: Antichain<T>) -> Self {
+        Self {
+            read_frontier: as_of.clone(),
+            write_frontier: as_of,
         }
     }
-}
 
-impl<T> Drop for CollectionIntrospection<T> {
-    fn drop(&mut self) {
-        self.refresh_introspection_state
-            .as_ref()
-            .map(|refresh_introspection_state| {
-                let retraction = refresh_introspection_state.row_for_collection(self.collection_id);
-                self.send(
-                    IntrospectionType::ComputeMaterializedViewRefreshes,
-                    vec![(retraction, -1)],
-                );
-            });
+    /// Return a `Row` reflecting the current collection frontiers.
+    fn row_for_collection(&self, collection_id: GlobalId) -> Row {
+        let read_frontier = self
+            .read_frontier
+            .as_option()
+            .map_or(Datum::Null, |ts| ts.clone().into());
+        let write_frontier = self
+            .write_frontier
+            .as_option()
+            .map_or(Datum::Null, |ts| ts.clone().into());
+        Row::pack_slice(&[
+            Datum::String(&collection_id.to_string()),
+            read_frontier,
+            write_frontier,
+        ])
+    }
+
+    /// Update the introspection state with the given new frontiers.
+    fn update(&mut self, read_frontier: AntichainRef<T>, write_frontier: AntichainRef<T>) {
+        if read_frontier != self.read_frontier.borrow() {
+            self.read_frontier = read_frontier.to_owned();
+        }
+        if write_frontier != self.write_frontier.borrow() {
+            self.write_frontier = write_frontier.to_owned();
+        }
```

Comment:
Should this protect against frontier regressions?

Reply:
I thought a bit about this and I think I like it better if the introspection types are "dumb" in the sense that they just propagate whatever frontier data is given to them. Frontier regressions don't cause correctness issues here and they are already checked and avoided by the calling code.

You could even imagine cases where we want to allow frontiers to regress in introspection. For example, we could reuse `ReplicaCollectionIntrospection` when we reconnect to a replica, in which case the write frontier would reset to a smaller value.

We're not doing any of this now though, so I'd be fine with checking for regressions if someone has a stronger opinion than me. It would make the code a bit more boilerplate-y, but I don't see any other drawbacks.